### DISCUSSION TITLE
How zero copy handles variable length data under same topic

### DISCUSSION THREAD

**Discussion by xixioba** [eProsima/Fast-DDS#2773] [Q&A]
Is there an example where this can be achieved :)

**Comment by JLBuenoLopez** [comment]
@xixioba,

Zero copy has some constraints one of them being that the type must be bounded (so no variable length is currently allowed). You can find more information in the corresponding section of [Fast DDS documentation](https://fast-dds.docs.eprosima.com/en/latest/fastdds/use_cases/zero_copy/zero_copy.html#constraints)

**Reply by xixioba** [reply]
Are there some plans to implement this feature?

**Reply by JLBuenoLopez** [reply]
It is not currently in the Fast DDS roadmap. You might contact our [support team](https://github.com/eProsima/Fast-DDS#getting-help) if you are interested in a feature acceleration commercial estimation.

**Comment by ZhenshengLee** [comment]
@xixioba Hi

this [ros2 design article of zero-copy](https://design.ros2.org/articles/zero_copy.html) talked about loaned_api with non-POD messages, which might help you.

**Reply by xixioba** [reply]
Thanks, 
I've looked at your repo: [ros2_shm_msgs](https://github.com/ZhenshengLee/ros2_shm_msgs), 
pre-defining data of various sizes (1m~8m) would cause a huge waste for my application, 
I'd love to know if there is a direction to fix this

**Reply by ZhenshengLee** [reply]
Oh, I didn't realize that, would you mind creating an issue in the repo of ros2_shm_msgs to describe your application requirement?
Or go to the discourse page in https://discourse.ros.org/t/using-zero-copy-transport-in-ros2-with-ros2-shm-msgs/26226  
Community will help!

Thanks.

**Reply by ZhenshengLee** [reply]
Although I don't know your application requirement.
I guess there will be trade off with mem waste, but will get benefit from latency decreasement. All depends on your needs and implementations.

**Reply by xixioba** [reply]
To put it simply, I need to transfer data such as protobuf, but I cannot predict the length (or maximum length) of the data

**Reply by ZhenshengLee** [reply]
In protobuf case, you could refer to https://github.com/continental/ecal/discussions/660 for the pitfalls of protobuf.

**Reply by xixioba** [reply]
Ok, I understand that protobuf serialization and deserialization take a considerable amount of time, maybe I will try capnproto, in fact I completed the test comparison through fastdds, capnp is indeed much faster, but unfortunately, I have to put the data in idl fixed size structure to enable zero copy, it can only be used for testing

**Reply by ZhenshengLee** [reply]
Technically, there is a design.

- For inter-machine-comm, use protobuf, for example gRPC in application layer.
- (In the edge between inter-machine and intra-machine)Create a bridge between protobuf and rosidl, for example https://github.com/ipa-rar/gRPC-ros-bridge , note that the bridge will add many times of copy, so the bridge must not be in the **Critical Chain** of the latency propagation chain.
- For intra-machine-comm, use fixed-size-idl with zero-copy to ensure the latency meeting performance requirement

**Reply by xixioba** [reply]
yes，I understand that this is the currently achievable solution
Also, I looked at [zenoh](https://github.com/eclipse-zenoh/zenoh) implementation and it seems that it can meet some of my needs [zero copy](https://github.com/eclipse-zenoh/zenoh/issues/46) and even communicate with existing [DDS](https://zenoh.io/blog/2021-04-28-ros2-integration/), I also see that you are involved, what do you say about zenoh's latency and load

**Reply by ZhenshengLee** [reply]
I didn't look it indepth, but I know there is a project from AutoCore(based on Nanjing, China) that would be good materials which is https://github.com/autocore-ai/zenoh_flow_autoware.

BTW, would you mind contacting me via wechat: zhensheng_li so we can discuss it more with Chinese community?

**Reply by xixioba** [reply]
ok, looks cool

**Comment by ZhenshengLee** [comment]
Also, the mechanism of zero copy of eCAL in https://continental.github.io/ecal/advanced/layers/shm.html#zero-copy-mode-optional does not require the bounded-size data types. Which may help.

**Reply by xixioba** [reply]
Thanks, I also tested ecal, unfortunately, it's not fully compatible with macos, even the pub/sub example doesn't work, ref https://github.com/continental/ecal/issues/697#issuecomment-1170668403, I'll try it again

**Comment by ZhenshengLee** [comment]
this issue https://github.com/eclipse-iceoryx/iceoryx/issues/911  describe an idea of dynamic size type support in iceoryx, which may help, but this has not been implemented yet.

**Reply by xixioba** [reply]
Thanks, I'm interested in eclipse-iceoryx, but when I test cyclonedds with iceoryx (on macos), it sometimes crashes, since iox-roudi needs to be run separately and it's difficult for me to manage, I went over the related issue https://github.com/eclipse-iceoryx/iceoryx/issues/970, maybe it works Feel free to use iceoryx, if you have experience I would really appreciate it
